internal/databases/clickhouse/bundle.go

```diff
@@ -0,0 +1,109 @@
+package clickhouse
+
+import (
+	"archive/tar"
+	"fmt"
+	"os"
+
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/internal"
+	"github.com/wal-g/wal-g/internal/crypto"
+	"github.com/wal-g/wal-g/utility"
+
+	"github.com/pkg/errors"
+)
+
+type TarSizeError struct {
+	error
+}
+
+func newTarSizeError(packedFileSize, expectedSize int64) TarSizeError {
+	return TarSizeError{errors.Errorf("packed wrong numbers of bytes %d instead of %d", packedFileSize, expectedSize)}
+}
+
+func (err TarSizeError) Error() string {
+	return fmt.Sprintf(tracelog.GetErrorFormatter(), err.error)
+}
+
+type Bundle struct {
+	Directory string
+
+	TarBallComposer TarBallComposer
+	TarBallQueue    *internal.TarBallQueue
+
+	Crypter  crypto.Crypter
+	Timeline uint32
+
+	TarSizeThreshold int64
+}
+
+func NewBundle(directory string, crypter crypto.Crypter, tarSizeThreshold int64) *Bundle {
+	return &Bundle{
+		Directory:        directory,
+		Crypter:          crypter,
+		TarSizeThreshold: tarSizeThreshold,
+	}
+}
+
+func (bundle *Bundle) StartQueue(tarBallMaker internal.TarBallMaker) error {
+	bundle.TarBallQueue = internal.NewTarBallQueue(bundle.TarSizeThreshold, tarBallMaker)
+	return bundle.TarBallQueue.StartQueue()
+}
+
+func (bundle *Bundle) SetupComposer(composerMaker TarBallComposerMaker) (err error) {
+	tarBallComposer, err := composerMaker.Make(bundle)
+	if err != nil {
+		return err
+	}
+	bundle.TarBallComposer = tarBallComposer
+	return nil
+}
+
+func (bundle *Bundle) HandleWalkedFSObject(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		if os.IsNotExist(err) {
+			tracelog.WarningLogger.Println(path, " deleted during filepath walk")
+			return nil
+		}
+		return errors.Wrap(err, "HandleWalkedFSObject: walk failed")
+	}
+
+	if info.IsDir() {
+		return nil
```

Comment:
I am not sure if we should skip the directories, even if they are empty. So I suggest calling [AddHeader]( https://github.com/wal-g/wal-g/blob/0261341f03bafb48960dbb5835d2b921cbb92b14/internal/databases/postgres/regular_tar_ball_composer.go#L73) like [it is done here](https://github.com/wal-g/wal-g/blob/196c92016f2ffb11b261f2e152f35ab066cdd6fb/internal/databases/postgres/bundle.go#L298)